In [1]:
%reset
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scipy
from scipy import stats
from scipy.stats import norm
import statsmodels.api as sm

Once deleted, variables cannot be recovered. Proceed (y/[n])?  y


# Power and Sample Size

Statistical power is the probability of of a statistical test to correctly reject the null hypthesis and find a significant effect when an effect exisits.  

# Power Calculations

In [2]:
import statsmodels.stats.power as smp

Example from leture of one sample, 1-sided test
Signal to Noise ratio = 110 versus 100.  Find the power. \
Different to be found = mean 1- true mean = 110 -100 = 10. \
Standard deviation = s= 20.  \
n= 16 observations. \
alpha = 0.05

power = probability(-t (at dof=n-1, 1-alpha) + ES) \
ES = |difference|/standard deviation = 10/SEM = 10/20/sqrt(16) = 10/5=2 \
power = probability (-1.753+2) = probability (0.247) = 60%

for z test power = probability(-z (at 0.95) + ES) \
power = probability (-1.645 + 2) = probability (0.355) = 63%

In [3]:
stats.t.ppf(0.95,15)

1.7530503556925547

In [4]:
stats.t.cdf(0.247,15)

0.5958715597043723

In [5]:
stats.norm.ppf(0.95)

1.6448536269514722

In [6]:
stats.norm.cdf(0.355)

0.6387052043836872

for using the power functions in python, the effect size is the diff/std estimate. \
for a one sample indepndent z-test power calculation, state that the ratio is 0. \
The one-sided test here is for larger test alternative. 

In [7]:
smp.zt_ind_solve_power(effect_size=(10/20),alpha=0.05, nobs1=16,ratio=0,alternative='larger')

0.638760031312335

for using a one-sample t-test power calculation, alternative='larger' for one sample

In [8]:
smp.ttest_power(effect_size=(10/20), nobs=16, alpha=0.05, alternative='larger')

0.6040328683316007

# For the much more common test of 2 independent samples, use the 2 sample power calculation for independent means. 

For example:  Suppose 100 oral contraceptive users and 100 non-oral contraceptive users are recruited for a study on systolic blood pressure (SBP) and oral contraceptive use.  A true differene in mean SBP is anticipated of 5 mmHg.  How much power would the study have assuming that the variances are similar and are OC users s=15.34 and non-OC users s=18.23.

power = probability (-z (at sig. level 1 - alpha/2) + ES) 

ES = difference in sample means / SEM

If you assume equal variance between samples:  the pooled standard deviation is the appropriate estimate of the standard deviation.  Then the SEM = s(pooled)*sqrt(1/n1 +1/n2)

s(pooled) = sqrt ( ((99)15.34^2 +(99)18.23^2) / (100+100-2) = sqrt( (23296.24 + 32900.96) / 198) = 16.85

ES = 5/(16.85* sqrt((1/100+1/100))) = 5/ (2.383) = 2.098

power= probability (z<-1.96 + 2.098) = (z<0.138) = 0.5557 = 56%

In [9]:
stats.norm.cdf(.138)

0.5548797912262711

In [10]:
# 2 sample t test power calculation
# ratio=1 for equal sample sizes
smp.tt_ind_solve_power(effect_size=(5/16.85),nobs1=100,alpha=0.05,alternative='two-sided',ratio=1)

0.5509876916375547

In [11]:
smp.zt_ind_solve_power(effect_size=(5/16.85),alpha=0.05, nobs1=100,ratio=1,alternative='two-sided')

0.5550134698036001

# Sample Size Calculations for 2-sample independent means test. 
n = 2* standard deviation ^2 * (z at 1-alpha + z at 1-beta)^2 / (mean1-mean2)^2

For the exmaple in lecture:  the sample size calculation for the one-sample, one sample t-test with: \
mean0= 100; mean1=110; difference in mean =-10 \
s= 20; alpha = 1-0.05 (one sided); beta = 0.1 for a power of 1-beta=0.90. \
 
z at 0.95 = 1.64 \
z at 0.90 = 1.29

nobs (using z distribution) = 20^2 *(1.64+1.29)^2 / 100 = 4 * 8.5849= 34.340
round up to the next interger to ensure adequate sample size

if you are worried about low numbers then we would use this n=35 for the sample size and use this in the t-distirbution to find a better sample size.

t at 0.95 and dof=34 = 1.69 \
t at 0.90 and dof=34 = 1.31

nobs= 20^2 * (1.69+1.31)^2 / (-10^2) = 400* (9) / 100 = 36 in the sample

In [12]:
stats.norm.ppf(0.95)

1.6448536269514722

In [13]:
stats.norm.ppf(0.90)

1.2815515655446004

In [14]:
smp.zt_ind_solve_power(effect_size=(10/20),alpha=0.05, power=0.9,ratio=0,alternative='larger')

34.25538932669148

In [15]:
stats.t.ppf(0.95,34)

1.6909242507706543

In [16]:
stats.t.ppf(0.90,34)

1.3069515856238303

# More common case is with an independent two-sample test of means.  For the example in lecture: 

estimate of difference=5 

standard deviation = 12, assume equal variance 

alpha =0.05 

beta= 0.20, power = 0.8 

z at 1-aplha/2 for 2-sided test = 1.96 

z at 0.80 = 0.84

n = 2* 12^2 * (1.96+0.84)^2 / 5^2     
  = 2 * 144 * (2.8)^2 / 25  = 90.317 or 91 per sample group.    

In [17]:
smp.tt_ind_solve_power(effect_size=(5/12),power=0.8,alpha=0.05,alternative='two-sided',ratio=1)

91.38922486804839

In [18]:
stats.norm.ppf(.975)

1.959963984540054

In [19]:
stats.norm.ppf(0.80)

0.8416212335729143

In [20]:
smp.zt_ind_solve_power(effect_size=(5/12),alpha=0.05,power=0.8,ratio=1,alternative='two-sided')

90.41887362430442